Created @scriptural/react package

abelpz opened the pull request:
Created lexical editor package for supporting usj

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/eten-tech-foundation/scripture-editors/221)
<!-- Reviewable:end -->

abelpz committed:
reuse nodes menu in platform


# Conflicts:
#	pnpm-lock.yaml

irahopkinson committed:
refactor making `getMakerAction` different per editor

irahopkinson committed:
get USJ marker action

irahopkinson committed:
Merge branch 'main' into abelpz-reuse-nodes-menu

irahopkinson committed:
Fix some things

- update lock file
- fix USJ version from `usfm2Usj`
- revert unintentional changes
- move nodes menu CSS
- fix menu closing when typing (improve selection change detection)
- pass in `scrRef` to plugin
- get `contextMarker`
- add poetry quote marker overwrites

abelpz committed:
added alignment support

abelpz committed:
added alignment support

abelpz committed:
Merge branch 'abelpz-reuse-nodes-menu' of https://github.com/BiblioNexus-Foundation/scripture-editors into scriptural-react

abelpz committed:
updae lock file

abelpz committed:
updae lock file

abelpz committed:
new-scriptural-react-package

abelpz committed:
new-scriptural-react-package

abelpz committed:
added docs

abelpz committed:
added docs

abelpz committed:
fix field-test found issues

abelpz committed:
fix field-test found issues

irahopkinson committed:
fix menu location on empty paragraph

- now at cursor rather than top-left

irahopkinson committed:
platform: add menu trigger property

irahopkinson committed:
platform fix: BCV when c & v inserted

abelpz committed:
fix scriptural marker nodes bugs with footnotes

abelpz committed:
fix scriptural marker nodes bugs with footnotes

abelpz committed:
update package-lock

abelpz committed:
fix css

abelpz committed:
removed debugger

abelpz committed:
fix relative path error

abelpz committed:
removed unused imports

abelpz committed:
delete schemas folder

abelpz committed:
removed comments

abelpz committed:
patch tests

abelpz committed:
fix cursor-handler bug with toggable notes

abelpz committed:
remove leftover console logs

abelpz committed:
replaced scriptural-app code with new playground app

abelpz committed:
make scriptureReferencePlugin update on verse of chapter update

abelpz committed:
improved scriptural-app and examples of persistence

abelpz committed:
fix css style format issues

abelpz committed:
downgraded esling config to traditional format

abelpz committed:
fix build warnings

abelpz committed:
further modularize toolbar buttons

abelpz committed:
removed unused import

abelpz committed:
fixed tailwind config for scriptural-app

abelpz committed:
added persistance examples to scriptural app

abelpz committed:
fixed extra whitespaces being added in usj to lexical conversion

irahopkinson reviewed (commented):
I'm publishing this partial review for now since GH is acting weird.

irahopkinson commented on packages/scriptural-react/CONTRIBUTING.md:
```diff
@@ -0,0 +1,306 @@
+
+3. Create a release PR:
+   - Update version in package.json
+   - Update CHANGELOG.md
```
The `CHANGELOG.md` file doesn't yet exist.

irahopkinson commented on packages/scriptural-react/package.json:
```diff
@@ -0,0 +1,60 @@
+  "peerDependencies": {
+    "react": ">=18.3.1",
+    "react-dom": ">=18.3.1",
+    "lexical": "^0.21.0",
+    "@lexical/react": "^0.21.0"
```
Maybe I'm wrong but aren't both of these lexical items actually `dependencies` rather than `peerDependencies`?

abelpz replied:
Not for this version of the package. Since I am allowing the user to add their own lexical extensions if needed, and they will probably need some lexical Types or functions that would not be bundled with this package. First it was a dependency but had to change it because of the above.

irahopkinson commented on packages/scriptural-react/package.json:
```diff
@@ -0,0 +1,60 @@
+  "author": {
+    "name": "BiblioNexus-Foundation",
+    "url": "https://github.com/BiblioNexus-Foundation"
+  },
```
Would you mind adding in volta to keep things consistent:
```json
  "volta": {
    "extends": "../../package.json"
  }
```

irahopkinson commented on apps/scriptural/package.json:
```diff
@@ -0,0 +1,37 @@
+    "@types/react-dom": "^18.3.5",
+    "globals": "^15.14.0",
+    "typescript-eslint": "^8.19.1"
+  }
```
Would you mind adding in volta to keep things consistent:
```json
  "volta": {
    "extends": "../../package.json"
  }
```

irahopkinson reviewed (commented):
Next partial review (GH still acting weird - perhaps its is the large number of files)

irahopkinson commented on packages/scriptural-react/src/plugins/ScripturalNodesMenuPlugin/useScripturalMarkersForMenu.ts:
```diff
@@ -0,0 +1,61 @@
+
+    if (!markerChildren || !marker) return;
+
+    return [...new Set([contextMarker, ...markerChildren, ...(markerAlike || [])])]
```
NIT use null coalescing operator instead, i.e. `markerAlike ?? []`.

irahopkinson commented on packages/scriptural-react/vite.config.ts:
I would appreciate a run through of this file with you.

abelpz replied:
Sure, I can meet at the same time as usual, anyday you can.

irahopkinson reviewed (commented):
Another partial review. I've run out of week now. I'll get back to this when I get back after next week. Sorry for the delay.

abelpz commented on apps/scriptural/src/App.tsx:
```diff
@@ -0,0 +1,152 @@
+        <PanelResizeHandle className="resizable-handle" />
+        <Panel defaultSize={50}>
+          <Editor
+            usj={editorState ? undefined : (getInitialUsj("TIT") ?? createEmptyUsj("TIT"))}
```
Yes it could be, it the cache was cleaned or entering for the first time. This is an example app of how the editor library could be used.

abelpz commented on apps/scriptural/src/App.tsx:
```diff
@@ -0,0 +1,152 @@
+    try {
+      const cached = JSON.parse(savedState) as CachedData<EditorState>;
+      console.log("Loading editor state from:", new Date(cached.timestamp).toISOString());
+      return JSON.stringify(cached.data);
```
I might remove all of this cache expiring thing, this was just the AI agent trying to add more stuff to the example.

abelpz commented on packages/scriptural-react/CONTRIBUTING.md:
````diff
@@ -0,0 +1,306 @@
+   ```
+3. Add the upstream repository:
+   ```bash
+   git remote add upstream https://github.com/bible-technology/scripture-editors.git
````
whoops, AI did a great Job with documentation, so good that I trusted it wouldn't fail with this.

abelpz commented on packages/scriptural-react/src/plugins/ChapterVerseUpdatePlugin/index.tsx:
```diff
@@ -0,0 +1,52 @@
+  const { scriptureReference } = useScripturalComposerContext();
+
+  useEffect(() => {
+    editor.registerNodeTransform(TextNode, (node) => {
```
You are right, some times I forget to do it, mostly if I am on a deadline. Thanks for catching that.

abelpz commented on packages/scriptural-react/src/plugins/ChapterVerseUpdatePlugin/index.tsx:
```diff
@@ -0,0 +1,52 @@
+  const { scriptureReference } = useScripturalComposerContext();
+
+  useEffect(() => {
+    editor.registerNodeTransform(TextNode, (node) => {
```
Good idea, thanks for suggesting it, will do.

abelpz commented on packages/shared/converters/usj/core/usj.d.ts:
Thanks!

abelpz committed:
fix PR review comments

irahopkinson commented on packages/shared/nodes/usfm/UsfmElementNode.ts:
```diff
@@ -2,7 +2,7 @@ import { ElementNode, LexicalNode, NodeKey, SerializedElementNode, Spread } from
 export type Attributes = { [key: string]: string };
 
-export type SerializedUsfmElementNode = Spread<
+export type SerializedScriptureElementNode = Spread<
```
Are you sure this should be renamed? It now clashes with another type of that same name. I would revert the 2 changes to this file if its not intentional.

irahopkinson commented on packages/shared/utils/scriptureMarkers/getScripturalMarkerActions.ts:
```diff
@@ -0,0 +1,315 @@
+    }
+  };
+
+  let nodeType: UsjNode["type"] | undefined = undefined;
```
NIT: assigning to `undefined` isn't needed since declaring it makes it undefined.

irahopkinson reviewed (changes requested):
Finally getting back to this review. It would be good to talk about your USJ nodes and if we can combine our usage of nodes.

abelpz committed:
Implement scripture reference handling and navigation in the scriptural app

- Introduced `AppReferenceHandler` to manage scripture references across editors.
- Enhanced `Editor` and `AppBar` components to utilize the new reference handler.
- Added `ScriptureNavigator` component for user-friendly navigation between scripture references.
- Integrated `ScrollToReferencePlugin` to automatically scroll to the current reference in the editor.
- Updated context and props to support scripture reference handling throughout the application.

abelpz committed:
Implement scripture reference handling and navigation in the scriptural editor lib and app (#301)

abelpz committed:
Remove unused import of `useLexicalComposerContext` from `ScrollToReferencePlugin` to clean up the code.

abelpz committed:
Refactor imports in ScriptureNavigator component to remove unused React import

abelpz committed:
Enhance ScrollToReferencePlugin to find verses within chapter context

- Added `findVerseInChapterContext` function to locate verses considering their chapter context.
- Implemented helper functions `isElementAfter` and `isElementBefore` to determine element positioning in the document.
- Updated `scrollToReference` to utilize the new verse-finding logic.

abelpz committed:
Implement Find and Replace functionality in the scriptural editor

- Added `FindReplacePlugin` to provide find and replace capabilities.
- Introduced `FindReplaceDialog` component for user interface.
- Integrated `SearchButton` in `CustomToolbar` to toggle the find and replace dialog.
- Updated `Editor` and related components to support the new functionality.
- Enhanced context management for search and replace operations.

abelpz committed:
Refactor imports in FindReplaceDialog component to remove unused React import

abelpz committed:
Enhance CustomToolbar and Editor components with trigger key functionality

- Added `TriggerKeyButton` to `CustomToolbar` for setting trigger key combinations.
- Integrated `TriggerKeyDialog` for user interaction with trigger keys.
- Updated `CustomToolbar` to include `ScripturalNodesMenuPlugin` and `CustomMarkersToolbar`.
- Wrapped `CustomToolbar` in `MarkersMenuProvider` within the `Editor` component for improved context management.
- Refactored toolbar layout for better organization and usability.

abelpz committed:
Refactor imports in CustomToolbar and TriggerKeyDialog components to remove unused imports

abelpz committed:
Remove unused option selection handler from NodesMenu component to streamline functionality

abelpz committed:
Remove unused ScripturalNodesMenuPlugin from CustomToolbar to streamline component functionality

abelpz committed:
Remove unused import of ScripturalNodesMenuPlugin from CustomToolbar to improve code clarity

irahopkinson commented on packages/shared/converters/usj/core/usj.d.ts:
```diff
@@ -0,0 +1,166 @@
+  content?: UsjParaContent[];
+  "link-id"?: NonWhitespaceString;
+  "link-href"?: NonWhitespaceString;
+  [key: `x-${string}`]: string;
```
I checked with the head of the USX/J/FM committee and it should be both `x-` and `z-` for properties. He helped me understand more about this if you want to chat in more detail.

abelpz committed:
Enhance save functionality and unsaved changes tracking in the scriptural editor

- Introduced `CustomSaveButton` in `CustomToolbar` to replace the previous `SaveButton`, adding support for unsaved changes indication.
- Implemented `useUnsavedChanges` hook in the `Editor` component to manage unsaved changes and integrate with the `HistoryPlugin`.
- Updated `SaveButton` to include automatic change detection and visual feedback for unsaved changes.
- Enhanced documentation to reflect new features and usage examples for tracking unsaved changes.

abelpz committed:
Update version in package.json for @scriptural/react to 0.0.22

irahopkinson commented:
Closing for now since no recent progress.
